File: JVM/univschedule.client/src/main/java/mntone/univschedule/client/UniversityScheduleClient.java
package mntone.univschedule.client;

import mntone.univschedule.core.*;

/**
 * Client for university class schedule
 */
public final class UniversityScheduleClient
{
	static final class UniversitiesBridge extends Universities.FriendUniversities
	{
	}

	static final class ClassesBridge extends Classes.FriendClasses
	{
	}

	static final UniversitiesBridge UNIVERSITIES_BRIDGE = new UniversitiesBridge();
	static final ClassesBridge CLASSES_BRIDGE = new ClassesBridge();

	static final String DEFAULT_USER_AGENT = "UniversityScheduleClient for Java/0.9";

	private final String mAccessKey;
	private String mAdditionalUserAgent = null;

	/**
	 * Initialize a new client for university class schedule.
	 *
	 * @param accessKey the access key
	 */
	public UniversityScheduleClient( String accessKey )
	{
		this.mAccessKey = accessKey;
	}

	/**
	 * Get universities.
	 *
	 * @return the array of University
	 * @throws UniversityScheduleException the UniversityClassScheduleException
	 */
	public Universities getUniversities() throws UniversityScheduleException
	{
		return UniversitiesClient.getUniversities( this );
	}

	/**
	 * Get classes.
	 *
	 * @param university the university
	 * @return the classes
	 * @throws UniversityScheduleException the UniversityClassScheduleException
	 */
	public Classes getClasses( University university ) throws UniversityScheduleException
	{
		return this.getClasses( university.getScreenName() );
	}

	/**
	 * Get classes.
	 *
	 * @param universityScreenName the screen name of university
	 * @return the classes
	 * @throws UniversityScheduleException the UniversityClassScheduleException
	 */
	public Classes getClasses( String universityScreenName ) throws UniversityScheduleException
	{
		return ClassesClient.getClasses( this, universityScreenName );
	}

	/**
	 * Get access key.
	 *
	 * @return the access key
	 */
	public String getAccessKey()
	{
		return this.mAccessKey;
	}

	/**
	 * Get additional user agent.
	 *
	 * @return the additional user agent
	 */
	public String getAdditionalUserAgent()
	{
		return this.mAdditionalUserAgent;
	}

	/**
	 * Set additional user agent.
	 *
	 * @param additionalUserAgent the additional user agent
	 */
	public void setAdditionalUserAgent( String additionalUserAgent )
	{
		this.mAdditionalUserAgent = additionalUserAgent;
	}
}

File: NET/UniversitySchedule.Client/UniversityScheduleClient.cs
using Mntone.UniversitySchedule.Client.Internal;
using Mntone.UniversitySchedule.Core;
using System;
using System.Net;
using System.Net.Http;
using System.Threading.Tasks;

namespace Mntone.UniversitySchedule.Client
{
	/// <summary>
	/// Client for university class schedule
	/// </summary>
	public sealed class UniversityScheduleClient
		: IDisposable
	{
		internal const string XUnivsLastModified = "X-UNIVS-LAST-MODIFIED";
		internal const string DefaultUserAgent = "UniversityScheduleClient for .NET/0.9";

		private HttpClientHandler _httpClientHandler = null;
		private HttpClient _httpClient = null;

		/// <summary>
		/// Initialize a new client for university class schedule.
		/// </summary>
		/// <param name="accessKey">Access key</param>
		public UniversityScheduleClient( string accessKey )
		{
			this._AccessKey = accessKey;
			this.AdditionalUserAgent = null;
			this.UniverstiesModifiedAt = DateTime.MinValue;
		}

		/// <summary>
		/// Dispose
		/// </summary>
		public void Dispose()
		{
			if( this._httpClient != null )
			{
				if( this._httpClientHandler != null )
				{
					this._httpClientHandler.Dispose();
					this._httpClientHandler = null;
				}

				this._httpClient.Dispose();
				this._httpClient = null;
			}
		}

		/// <summary>
		/// Get universities.
		/// </summary>
		/// <returns>The array of <see cref="University"/></returns>
		public Task<UniversitiesResponse> GetUniversitiesAsync()
		{
			var url = string.Format( UniversityScheduleUrls.UNIVERSTIES_URL, this.AccessKey );
			return new UniversitiesClient().GetAsync( this, url );
		}

		/// <summary>
		/// Get classes.
		/// </summary>
		/// <returns>The array of <see cref="Class"/></returns>
		public Task<ClassesResponse> GetClassesAsync( University university )
		{
			return GetClassesAsync( university.ScreenName );
		}

		/// <summary>
		/// Get classes.
		/// </summary>
		/// <returns>The array of <see cref="Class"/></returns>
		public Task<ClassesResponse> GetClassesAsync( string universityScreenName )
		{
			var url = string.Format( UniversityScheduleUrls.CANCELLATIONS_URL, this.AccessKey, universityScreenName );
			return new BaseClient<ClassesResponse>().GetAsync( this, url );
		}

		internal HttpClient GetClient()
		{
			if( this._httpClient == null )
			{
				this._httpClientHandler = new HttpClientHandler();
				this._httpClientHandler.AllowAutoRedirect = false;
				this._httpClientHandler.AutomaticDecompression = DecompressionMethods.Deflate | DecompressionMethods.GZip;
				this._httpClient = new HttpClient( this._httpClientHandler, false );
				this._httpClient.DefaultRequestHeaders.Add(
					"user-agent",
					this.AdditionalUserAgent != null
						? DefaultUserAgent + " (" + this.AdditionalUserAgent + ')'
						: DefaultUserAgent );
				this._httpClient.Timeout = TimeSpan.FromSeconds( 5 );
			}
			return this._httpClient;
		}

		/// <summary>
		/// Access key
		/// </summary>
		public string AccessKey
		{
			get { return this._AccessKey; }
		}
		private readonly string _AccessKey = null;

		/// <summary>
		/// Additional user agent
		/// </summary>
		public string AdditionalUserAgent { get; set; }

		/// <summary>
		/// Last modified of university list
		/// </summary>
		public DateTime UniverstiesModifiedAt { get; set; }
	}
}